File: search2d.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        low,high=0,len(matrix)*len(matrix[0])-1
        
        
        while low<=high:
            mid=(low+high)//2
            row=mid//len(matrix[0])
            col=mid%len(matrix[0])
            
            if matrix[row][col]==target:
                return True
            elif matrix[row][col]>target:
                high=mid-1
            else:
                low=mid+1
        return False
            

File: search2d.cpp

// TC:O(log(m*n))
// SC:O(1)

// Approach:
// Idea is flatten the 2D array and then perform the search by getting the row and col with 
// row = val/no.of.columns; col = val/no.of.columns


class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {

        int r=matrix.size();
        int c=matrix[0].size();

         int low=0;
         int high=r*c-1;

         while(low <= high)
         {
             int mid= low+(high-low)/2;
             
             if(matrix[mid/c][mid%c]==target)
             {
                 return true;
             }
             else if(matrix[mid/c][mid%c]<target)
             {
                 low=mid+1;
             }
             else
             {
                 high=mid-1;
             }
         }
        return false;
    }
};